include/deal.II/fe/fe_interface.h

```diff
@@ -0,0 +1,787 @@
+      Assert(fe_index == 0,
+             ExcMessage("A boundary facet only has FE index 0."));
+
+    if (fe_index == 0)
+      {
+        Assert(face_dof_index < n_dofs_fe, ExcMessage("invalid face_dof_idx"));
+        return face_dof_index;
+      }
+    else if (fe_index == 1)
+      {
+        Assert(face_dof_index < n_dofs_fe_neighbor,
+               ExcMessage("invalid face_dof_idx"));
+        return face_dof_index + n_dofs_fe;
+      }
+    // return something invalid:
+    return -1;
+  }
+
+  /**
+   * Return the FEFaceValue object of the cell or the neighboring cell
+   */
+  const FEFaceValuesBase<dim, spacedim> &
+  get_fe_values(const unsigned int cell_or_neighbor) const
+  {
+    AssertIndexRange(cell_or_neighbor, 2);
+    return (cell_or_neighbor == 0) ? get_fe_values() : get_fe_values_neighbor();
+  }
+
+  const FEFaceValuesBase<dim, spacedim> &
+  get_fe_values() const
+  {
+    return *fe_face_values;
+  }
+
+  const FEFaceValuesBase<dim, spacedim> &
+  get_fe_values_neighbor() const
+  {
+    Assert(!is_boundary_facet(),
+           ExcMessage("Not possible for boundary Facet."));
+    return *fe_face_values_neighbor;
+  }
+
+  /**
+   * Return the normal in a given quadrature point.
+   */
+  Tensor<1, spacedim>
+  normal(const unsigned int q_point_index) const
+  {
+    return fe_face_values->normal_vector(q_point_index);
+  }
+
+  /**
+   * Return a view of scalar FE.
+   */
+  const FEInterfaceViews::Scalar<dim, spacedim> &
+  operator[](const FEValuesExtractors::Scalar &scalar) const
+  {
+    Assert(scalar.component < fe_face_values->get_fe().n_components(),
+           ExcMessage("Invalid FEValuesExtractors::Scalar!"));
+
+    return cached_views_scalar[scalar.component];
+  }
+
+  /**
+   * Return *this[FEValuesExtractors::Scalar(0)] if this FE is scalar (has
+   * only a single component).
+   *
+   * Fail otherwise.
+   */
+  const FEInterfaceViews::Scalar<dim, spacedim> &
+  scalar() const
+  {
+    Assert(
+      fe_face_values->get_fe().n_components() == 1,
+      ExcMessage(
+        "FEInterfaceValues::scalar() is only valid if the FE has exactly one component!"));
+    return cached_views_scalar[0];
+  }
+
+  /**
+   * Create a view of the current FEValues object that represents a set of
+   * <code>dim</code> scalar components (i.e. a vector) of the vector-valued
+   * finite element. The concept of views is explained in the documentation of
+   * the namespace FEValuesViews and in particular in the
+   * @ref vector_valued
+   * module.
+   */
+  const FEInterfaceViews::Vector<dim, spacedim> &
+  operator[](const FEValuesExtractors::Vector &vector) const
+  {
+    Assert(vector.first_vector_component + spacedim <=
+             fe_face_values->get_fe().n_components(),
+           ExcMessage("Invalid FEValuesExtractors::Vector!"));
+
+    return cached_views_vector[vector.first_vector_component];
+  }
+
+  /**
+   * Return the left or the right (if @p left is false) value on the interface for the shape
```

Comment:
It is unclear here that left means *this cell* and right means *the other cell*. I would not assume this, and rename the argument (and the documentation) to say `current_cell` (or neighbor cell).